File: app/src/test/java/network/spiritscorp/InsulinRepositoryIntegrationTest.java
package network.spiritscorp;

/*
 * Copyright (C) 2026 Tom Spirit
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import androidx.annotation.NonNull;
import kotlin.coroutines.EmptyCoroutineContext;
import kotlinx.coroutines.BuildersKt;
import kotlinx.coroutines.flow.Flow;
import kotlinx.coroutines.flow.FlowKt;
import kotlinx.coroutines.flow.MutableStateFlow;
import kotlinx.coroutines.flow.StateFlowKt;
import network.spiritscorp.data.CalculationLogDao;
import network.spiritscorp.data.InsulinRepository;
import network.spiritscorp.data.UserSettingsDao;
import network.spiritscorp.model.CalculationLog;
import network.spiritscorp.model.UserSettings;
import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

/**
 * Java integration tests verifying the interaction between the data access layer,
 * reactive Flows, and the InsulinRepository domain coordinator.
 */
public class InsulinRepositoryIntegrationTest {

    private static final double DELTA = 0.001;

    private FakeUserSettingsDao fakeSettingsDao;
    private InsulinRepository repository;

    @Before
    public void setup() {
        FakeCalculationLogDao fakeLogDao = new FakeCalculationLogDao();
        fakeSettingsDao = new FakeUserSettingsDao();
        repository = new InsulinRepository(fakeLogDao, fakeSettingsDao);
    }

    @Test
    public void testDefaultSettingsCreatedWhenNoneExist() {
        UserSettings settings = repository.getSettings();
        assertNotNull(settings);
        assertEquals(1.50, settings.getMorningFactor(), DELTA);
        assertEquals("GRAMS", settings.getDefaultCarbUnit());
        assertEquals(120.0, settings.getTargetGlucoseMgDl(), DELTA);
        assertEquals(50.0, settings.getCorrectionFactorMgDl(), DELTA);

        // Verify default settings were persisted to DAO
        UserSettings direct = fakeSettingsDao.getSettingsDirect();
        assertNotNull(direct);
        assertEquals(1.50, direct.getMorningFactor(), DELTA);
    }

    @Test
    public void testSaveAndRetrieveCustomSettings() {
        UserSettings custom = new UserSettings(
                1,
                1.75,
                1.15,
                1.40,
                0.85,
                "BE",
                12,
                "mg/dl",
                105.0,
                45.0,
                0.5,
                true,
                "AMBER_WARM",
                "SYSTEM"
        );

        repository.saveSettings(custom);

        UserSettings retrieved = repository.getSettings();
        assertEquals(1.75, retrieved.getMorningFactor(), DELTA);
        assertEquals("BE", retrieved.getDefaultCarbUnit());
        assertEquals("AMBER_WARM", retrieved.getSelectedTheme());

        UserSettings flowValue = firstFromFlow(repository.getSettingsFlow());
        assertNotNull(flowValue);
        assertEquals("AMBER_WARM", flowValue.getSelectedTheme());
    }

    @Test
    public void testSaveSingleCalculationLogAndObserveFlow() {
        CalculationLog log = new CalculationLog(
                0L,
                System.currentTimeMillis(),
                "Mittagessen (Reis mit Hühnchen)",
                60.0,
                "g KH",
                60.0,
                5.0,
                6.0,
                "Mittags",
                1.0,
                5.0,
                125.0,
                100.0,
                40.0,
                0.63,
                5.63,
                5.5,
                "Leichte Sporteinheit danach"
        );

        long id = repository.saveCalculation(log);
        assertTrue(id > 0);

        List<CalculationLog> logsFromFlow = firstFromFlow(repository.getAllLogs());
        assertNotNull(logsFromFlow);
        assertEquals(1, logsFromFlow.size());
        assertEquals("Mittagessen (Reis mit Hühnchen)", logsFromFlow.getFirst().getMealTitle());
        assertEquals(5.5, logsFromFlow.getFirst().getRoundedInsulin(), DELTA);

        List<CalculationLog> directLogs = repository.getAllLogsDirect();
        assertEquals(1, directLogs.size());
    }

    @Test
    public void testSaveMultipleLogsBatchAndOrdering() {
        CalculationLog log1 = new CalculationLog(1L, 1000L, "Mahlzeit 1", 0.0, "g KH", 0.0, 0.0, 0.0, "Morgens", 1.0, 0.0, null, null, null, 0.0, 0.0, 0.0, "");
        CalculationLog log2 = new CalculationLog(2L, 2000L, "Mahlzeit 2", 0.0, "g KH", 0.0, 0.0, 0.0, "Mittags", 1.0, 0.0, null, null, null, 0.0, 0.0, 0.0, "");
        CalculationLog log3 = new CalculationLog(3L, 3000L, "Mahlzeit 3", 0.0, "g KH", 0.0, 0.0, 0.0, "Abends", 1.0, 0.0, null, null, null, 0.0, 0.0, 0.0, "");

        long[] insertedIds = repository.saveLogs(Arrays.asList(log1, log2, log3));
        assertEquals(3, insertedIds.length);

        List<CalculationLog> allLogs = repository.getAllLogsDirect();
        assertEquals(3, allLogs.size());
        // Verify chronological descending order
        assertEquals("Mahlzeit 3", allLogs.get(0).getMealTitle());
        assertEquals("Mahlzeit 2", allLogs.get(1).getMealTitle());
        assertEquals("Mahlzeit 1", allLogs.get(2).getMealTitle());
    }

    @Test
    public void testDeleteLogById() {
        CalculationLog log1 = new CalculationLog(10L, 1000L, "Frühstück", 0.0, "g KH", 0.0, 0.0, 0.0, "Morgens", 1.0, 0.0, null, null, null, 0.0, 0.0, 0.0, "");
        CalculationLog log2 = new CalculationLog(20L, 2000L, "Abendessen", 0.0, "g KH", 0.0, 0.0, 0.0, "Abends", 1.0, 0.0, null, null, null, 0.0, 0.0, 0.0, "");

        repository.saveLogs(Arrays.asList(log1, log2));
        List<CalculationLog> initialLogs = repository.getAllLogsDirect();
        assertEquals(2, initialLogs.size());

        repository.deleteLog(10L);
        List<CalculationLog> remaining = repository.getAllLogsDirect();
        assertEquals(1, remaining.size());
        assertEquals("Abendessen", remaining.getFirst().getMealTitle());
    }

    @Test
    public void testClearAllLogs() {
        CalculationLog log1 = new CalculationLog(1L, 1000L, "Eintrag 1", 0.0, "g KH", 0.0, 0.0, 0.0, "Morgens", 1.0, 0.0, null, null, null, 0.0, 0.0, 0.0, "");
        CalculationLog log2 = new CalculationLog(2L, 2000L, "Eintrag 2", 0.0, "g KH", 0.0, 0.0, 0.0, "Mittags", 1.0, 0.0, null, null, null, 0.0, 0.0, 0.0, "");
        repository.saveLogs(Arrays.asList(log1, log2));

        List<CalculationLog> initialLogs = repository.getAllLogsDirect();
        assertEquals(2, initialLogs.size());

        repository.clearLogs();
        List<CalculationLog> remainingLogs = repository.getAllLogsDirect();
        assertEquals(0, remainingLogs.size());
        List<CalculationLog> flowLogs = firstFromFlow(repository.getAllLogs());
        assertEquals(0, flowLogs.size());
    }

    // --- Coroutine Helper Methods for Flow collection in Java ---

    public static <T> T firstFromFlow(Flow<T> flow) {
        try {
            return BuildersKt.runBlocking(
                    EmptyCoroutineContext.INSTANCE,
                    (_, continuation) -> FlowKt.first(flow, continuation)
            );
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RuntimeException("Flow collection interrupted", e);
        }
    }

    // --- Fake DAO Implementations ---

    public static class FakeCalculationLogDao implements CalculationLogDao {
        private final List<CalculationLog> logs = new ArrayList<>();
        private final MutableStateFlow<List<CalculationLog>> flow = StateFlowKt.MutableStateFlow(Collections.emptyList());
        private long nextId = 1L;

        private synchronized void updateFlow() {
            List<CalculationLog> sorted = new ArrayList<>(logs);
            sorted.sort((a, b) -> Long.compare(b.getTimestamp(), a.getTimestamp()));
            flow.setValue(Collections.unmodifiableList(sorted));
        }

        @NonNull
        @Override
        public Flow<List<CalculationLog>> getAllLogs() {
            return flow;
        }

        @NonNull
        @Override
        public List<CalculationLog> getAllLogsDirect() {
            synchronized (this) {
                List<CalculationLog> sorted = new ArrayList<>(logs);
                sorted.sort((a, b) -> Long.compare(b.getTimestamp(), a.getTimestamp()));
                return sorted;
            }
        }

        @Override
        public long insertLog(@NonNull CalculationLog log) {
            synchronized (this) {
                long assignedId = log.getId() == 0L ? nextId++ : log.getId();
                CalculationLog toSave = log.copy();
                toSave.setId(assignedId);
                logs.removeIf(l -> l.getId() == assignedId);
                logs.add(toSave);
                updateFlow();
                return assignedId;
            }
        }

        @Override
        public long[] insertLogs(@NonNull List<CalculationLog> newLogs) {
            synchronized (this) {
                long[] ids = new long[newLogs.size()];
                for (int i = 0; i < newLogs.size(); i++) {
                    CalculationLog log = newLogs.get(i);
                    long assignedId = log.getId() == 0L ? nextId++ : log.getId();
                    CalculationLog toSave = log.copy();
                    toSave.setId(assignedId);
                    logs.removeIf(l -> l.getId() == assignedId);
                    logs.add(toSave);
                    ids[i] = assignedId;
                }
                updateFlow();
                return ids;
            }
        }

        @Override
        public void deleteLogById(long logId) {
            synchronized (this) {
                logs.removeIf(l -> l.getId() == logId);
                updateFlow();
            }
        }

        @Override
        public void clearAllLogs() {
            synchronized (this) {
                logs.clear();
                updateFlow();
            }
        }
    }

    public static class FakeUserSettingsDao implements UserSettingsDao {
        private UserSettings storedSettings = null;
        private final MutableStateFlow<UserSettings> flow = StateFlowKt.MutableStateFlow(null);

        @NonNull
        @Override
        public Flow<UserSettings> getSettings() {
            return flow;
        }

        @Override
        public UserSettings getSettingsDirect() {
            return storedSettings;
        }

        @Override
        public void saveSettings(@NonNull UserSettings settings) {
            storedSettings = settings;
            flow.setValue(settings);
        }
    }
}

File: app/src/test/java/network/spiritscorp/InsulinRepositoryIntegrationTest.kt
package network.spiritscorp

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.test.runTest
import network.spiritscorp.data.CalculationLogDao
import network.spiritscorp.data.InsulinRepository
import network.spiritscorp.data.UserSettingsDao
import network.spiritscorp.model.CalculationLog
import network.spiritscorp.model.UserSettings
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

/**
 * Integration tests verifying the interaction between the data access layer,
 * reactive Flows, suspend queries, and the InsulinRepository domain coordinator.
 */
class InsulinRepositoryIntegrationTest {

    private lateinit var fakeLogDao: FakeCalculationLogDao
    private lateinit var fakeSettingsDao: FakeUserSettingsDao
    private lateinit var repository: InsulinRepository

    @Before
    fun setup() {
        fakeLogDao = FakeCalculationLogDao()
        fakeSettingsDao = FakeUserSettingsDao()
        repository = InsulinRepository(fakeLogDao, fakeSettingsDao)
    }

    @Test
    fun testDefaultSettingsCreatedWhenNoneExist() = runTest {
        val settings = repository.getSettings()
        assertNotNull(settings)
        assertEquals(1.50, settings.morningFactor, 0.001)
        assertEquals("GRAMS", settings.defaultCarbUnit)
        assertEquals(120.0, settings.targetGlucoseMgDl, 0.001)
        assertEquals(50.0, settings.correctionFactorMgDl, 0.001)

        // Verify it was persisted
        val direct = fakeSettingsDao.getSettingsDirect()
        assertNotNull(direct)
        assertEquals(1.50, direct!!.morningFactor, 0.001)
    }

    @Test
    fun testSaveAndRetrieveCustomSettings() = runTest {
        val custom = UserSettings(
            morningFactor = 1.75,
            noonFactor = 1.15,
            eveningFactor = 1.40,
            nightFactor = 0.85,
            defaultCarbUnit = "BE",
            beGramsDivisor = 12,
            targetGlucoseMgDl = 105.0,
            correctionFactorMgDl = 45.0,
            roundingStep = 0.5,
            selectedTheme = "AMBER_WARM"
        )

        repository.saveSettings(custom)

        val retrieved = repository.getSettings()
        assertEquals(1.75, retrieved.morningFactor, 0.001)
        assertEquals("BE", retrieved.defaultCarbUnit)
        assertEquals("AMBER_WARM", retrieved.selectedTheme)

        val flowValue = repository.settingsFlow.first()
        assertEquals("AMBER_WARM", flowValue?.selectedTheme)
    }

    @Test
    fun testSaveSingleCalculationLogAndObserveFlow() = runTest {
        val log = CalculationLog(
            mealTitle = "Mittagessen (Reis mit Hühnchen)",
            rawCarbInput = 60.0,
            carbUnit = "g KH",
            carbGrams = 60.0,
            beValue = 5.0,
            keValue = 6.0,
            timeOfDay = "Mittags",
            insulinFactor = 1.0,
            mealInsulin = 5.0,
            bloodGlucose = 125.0,
            targetGlucose = 100.0,
            correctionFactor = 40.0,
            correctionInsulin = 0.63,
            totalInsulin = 5.63,
            roundedInsulin = 5.5,
            notes = "Leichte Sporteinheit danach"
        )

        val id = repository.saveCalculation(log)
        assertTrue(id > 0)

        val logsFromFlow = repository.allLogs.first()
        assertEquals(1, logsFromFlow.size)
        assertEquals("Mittagessen (Reis mit Hühnchen)", logsFromFlow[0].mealTitle)
        assertEquals(5.5, logsFromFlow[0].roundedInsulin, 0.001)

        val directLogs = repository.getAllLogsDirect()
        assertEquals(1, directLogs.size)
    }

    @Test
    fun testSaveMultipleLogsBatchAndOrdering() = runTest {
        val log1 = CalculationLog(id = 1L, timestamp = 1000L, mealTitle = "Mahlzeit 1")
        val log2 = CalculationLog(id = 2L, timestamp = 2000L, mealTitle = "Mahlzeit 2")
        val log3 = CalculationLog(id = 3L, timestamp = 3000L, mealTitle = "Mahlzeit 3")

        val insertedIds = repository.saveLogs(listOf(log1, log2, log3))
        assertEquals(3, insertedIds.size)

        val allLogs = repository.getAllLogsDirect()
        assertEquals(3, allLogs.size)
        // Check order descending by timestamp
        assertEquals("Mahlzeit 3", allLogs[0].mealTitle)
        assertEquals("Mahlzeit 2", allLogs[1].mealTitle)
        assertEquals("Mahlzeit 1", allLogs[2].mealTitle)
    }

    @Test
    fun testDeleteLogById() = runTest {
        val log1 = CalculationLog(id = 10L, timestamp = 1000L, mealTitle = "Frühstück")
        val log2 = CalculationLog(id = 20L, timestamp = 2000L, mealTitle = "Abendessen")

        repository.saveLogs(listOf(log1, log2))
        assertEquals(2, repository.getAllLogsDirect().size)

        repository.deleteLog(10L)
        val remaining = repository.getAllLogsDirect()
        assertEquals(1, remaining.size)
        assertEquals("Abendessen", remaining[0].mealTitle)
    }

    @Test
    fun testClearAllLogs() = runTest {
        val log1 = CalculationLog(id = 1L, timestamp = 1000L, mealTitle = "Eintrag 1")
        val log2 = CalculationLog(id = 2L, timestamp = 2000L, mealTitle = "Eintrag 2")
        repository.saveLogs(listOf(log1, log2))

        assertEquals(2, repository.getAllLogsDirect().size)

        repository.clearLogs()
        assertEquals(0, repository.getAllLogsDirect().size)
        assertEquals(0, repository.allLogs.first().size)
    }

    // --- Fake DAO Implementations ---

    private class FakeCalculationLogDao : CalculationLogDao {
        private val logs = mutableListOf<CalculationLog>()
        private val _flow = MutableStateFlow<List<CalculationLog>>(emptyList())
        private var nextId = 1L

        private fun updateFlow() {
            _flow.value = logs.sortedByDescending { it.timestamp }.toList()
        }

        override fun getAllLogs(): Flow<List<CalculationLog>> = _flow.asStateFlow()

        override suspend fun getAllLogsDirect(): List<CalculationLog> {
            return logs.sortedByDescending { it.timestamp }.toList()
        }

        override suspend fun insertLog(log: CalculationLog): Long {
            val assignedId = if (log.id == 0L) nextId++ else log.id
            val toSave = log.copy(id = assignedId)
            logs.removeAll { it.id == assignedId }
            logs.add(toSave)
            updateFlow()
            return assignedId
        }

        override suspend fun insertLogs(logs: List<CalculationLog>): List<Long> {
            val ids = mutableListOf<Long>()
            for (l in logs) {
                ids.add(insertLog(l))
            }
            return ids
        }

        override suspend fun deleteLogById(logId: Long) {
            logs.removeAll { it.id == logId }
            updateFlow()
        }

        override suspend fun clearAllLogs() {
            logs.clear()
            updateFlow()
        }
    }

    private class FakeUserSettingsDao : UserSettingsDao {
        private var storedSettings: UserSettings? = null
        private val _flow = MutableStateFlow<UserSettings?>(null)

        override fun getSettings(): Flow<UserSettings?> = _flow.asStateFlow()

        override suspend fun getSettingsDirect(): UserSettings? = storedSettings

        override suspend fun saveSettings(settings: UserSettings) {
            storedSettings = settings
            _flow.value = settings
        }
    }
}
